### ISSUE TITLE
dev和test数据集预测阶段，迭代时间很慢，一次迭代要2个多小时。请问作者这个问题您遇到过嘛，我怀疑会不会是您is_eval中内部循环的原因

### ISSUE THREAD

**Issue by lujiang001108** [neukg/MultiIE#2]
def forward(self, q_ids=None, passage_ids=None, segment_ids=None, attention_mask=None, subject_labels=None, object_labels=None, 
                event_num_labels=None,
                eval_file=None,
                is_eval=False):

        # bert_encoder(batch_size, seq_len, hidden_size)
        bert_encoder, pooler_out, hidden_states = self.encoder(passage_ids, attention_mask, segment_ids, output_hidden_states=True)
        # 1: trigger: last two  argument: last one
        trigger_represent = hidden_states[-2]
        argument_represent = bert_encoder
        # 2. trigger: dynamic fusion  argument: dynamic fusion
        #trigger_represent = self.dynamic_fusion(hidden_states=hidden_states[1:], weights=self.trigger_weight)
        #argument_represent = self.dynamic_fusion(hidden_states=hidden_states[1:], weights=self.argument_weight) 
        if not is_eval:
            seq_len = trigger_represent.size(1)
            start_sub = trigger_represent * subject_labels[:, :, 0].unsqueeze(2)
            start_sub = nn.AvgPool1d(seq_len)(start_sub.transpose(1, 2))
            start_sub = start_sub.transpose(1, 2)
            end_sub = trigger_represent * subject_labels[:, :, 1].unsqueeze(2)
            end_sub = nn.AvgPool1d(seq_len)(end_sub.transpose(1, 2))
            end_sub = end_sub.transpose(1, 2)

            # batch, 1, hidden_size * 2
            subject = torch.cat([start_sub, end_sub], 2)
            att_weight = self.additive_attention(argument_represent, subject, attention_mask).unsqueeze(2)
            context_encoder = argument_represent * att_weight
            context_encoder = torch.cat([context_encoder, argument_represent], dim=2)
            context_encoder = self.context_dense(context_encoder)

            sub_preds = nn.Sigmoid()(self.subject_dense(trigger_represent))
            po_logits = self.event_decoder(context_encoder, attention_mask)
            po_preds = nn.Sigmoid()(po_logits.reshape(passage_ids.size(0), passage_ids.size(1), self.event_num, self.classes_num, 2))

            subject_loss = self.loss_fct(sub_preds, subject_labels)
            subject_loss = subject_loss.mean(2)
            subject_loss = torch.sum(subject_loss * attention_mask.float()) / torch.sum(attention_mask.float())
            
            po_loss = self.bi_match_loss(po_preds, object_labels, attention_mask)

            loss = subject_loss + po_loss

            return loss

        else:
            # (batch_size, seq_len, 2)
            subject_preds = nn.Sigmoid()(self.subject_dense(trigger_represent))
            answer_list = list()
            subject_ids = []
            for qid, sub_pred, trigger_repre in zip(q_ids.cpu().numpy(),
                                     subject_preds.cpu().numpy(), trigger_represent.cpu().numpy()):
                print(q_ids)
                context = eval_file[qid].bert_tokens
                # start:(true_number, )
                start = np.where(sub_pred[:, 0] > config.eval_config["sub_threshold_start"])[0]
                end = np.where(sub_pred[:, 1] > config.eval_config["sub_threshold_end"])[0]
                if start.size == 0 or end.size == 0:
                    subject_ids.append((-1, -1))
                    answer_list.append(np.zeros((trigger_repre.shape[1] * 2)))
                    continue
                subject = []
                s_starts = []
                s_ends = []
                for i in start:
                    j = end[end >= i]
                    if i == 0 or i > len(context) - 2:
                        continue

                    if len(j) > 0:
                        j = j[0]
                        if j > len(context) - 2:
                            continue
                        subject.append((i, j))
                        s_starts.append(trigger_repre[i, :])
                        s_ends.append(trigger_repre[j, :])
                if not s_starts or not s_ends:
                    subject_ids.append((-1, -1))
                    answer_list.append(np.zeros((trigger_repre.shape[1] * 2)))
                    continue
                subject_ids.append(subject)
                s_starts = np.average(np.array(s_starts), axis=0)
                s_ends = np.average(np.array(s_ends), axis=0)
                answer_list.append(np.concatenate((s_starts, s_ends), axis=0))

            sub_tensor = torch.tensor(answer_list, dtype=torch.float32).unsqueeze(1).cuda()
            att_weight = self.additive_attention(argument_represent, sub_tensor, attention_mask).unsqueeze(2)
            context_encoder = argument_represent * att_weight
            context_encoder = torch.cat([context_encoder, argument_represent], dim=2)
            context_encoder = self.context_dense(context_encoder)
            po_logits = self.event_decoder(context_encoder, attention_mask)
            po_preds = nn.Sigmoid()(po_logits.reshape(passage_ids.size(0), passage_ids.size(1), self.event_num, self.classes_num, 2))
            return q_ids, po_preds
